using static Directive doesn't compile

Currently a script with [using static directive](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-static) like below

```cake
using static System.Math;

Information(Round(1.1));
```
will fail with
```diff
- Error: Error occurred when compiling build script: error CS0246: 
- The type or namespace name 'static' could not be found (are you missing a using directive or an assembly reference?)
- (3,13): error CS0103: The name 'Round' does not exist in the current context
```

As we've unified around Roslyn this can now and should be supported by Cake, I'll submit a pull request shortly with a proposed fix.

--- a/src/Cake.Core.Tests/Unit/Scripting/Analysis/ScriptAnalyzerResultTests.cs
+++ b/src/Cake.Core.Tests/Unit/Scripting/Analysis/ScriptAnalyzerResultTests.cs
@@ -134,6 +134,25 @@ namespace Cake.Core.Tests.Unit.Scripting.Analysis
                 // Then
                 Assert.True(result.Succeeded);
             }
+
+            [Fact]
+            public void Should_Populate_Using_Static_From_Provided_Script()
+            {
+                // Given
+                var script = new ScriptInformation("./build.cake");
+                script.UsingStaticDirectives.Add("using static System.Math;");
+                var other = new ScriptInformation("./other.cake");
+                other.UsingStaticDirectives.Add("using static System.Console;");
+                script.Includes.Add(other);
+
+                // When
+                var result = new ScriptAnalyzerResult(script, new List<string>());
+
+                // Then
+                Assert.Equal(2, result.UsingStaticDirectives.Count);
+                Assert.Contains("using static System.Math;", result.UsingStaticDirectives);
+                Assert.Contains("using static System.Console;", result.UsingStaticDirectives);
+            }
         }
     }
 }

--- a/src/Cake.Core.Tests/Unit/Scripting/Analysis/ScriptAnalyzerTests.cs
+++ b/src/Cake.Core.Tests/Unit/Scripting/Analysis/ScriptAnalyzerTests.cs
@@ -369,6 +369,21 @@ namespace Cake.Core.Tests.Unit.Scripting.Analysis
                 Assert.Equal(2, result.Errors[0].Line);
                 Assert.Equal("Query string for #load contains more than one parameter 'path'.", result.Errors[0].Message);
             }
+
+            [Fact]
+            public void Should_Process_Using_Static_Directives()
+            {
+                // Given
+                var fixture = new ScriptAnalyzerFixture();
+                fixture.GivenScriptExist("/Working/script.cake", "using static System.Math;");
+
+                // When
+                var result = fixture.Analyze("/Working/script.cake");
+
+                // Then
+                Assert.Equal(1, result.Script.UsingStaticDirectives.Count);
+                Assert.Equal("using static System.Math;", result.Script.UsingStaticDirectives.ElementAt(0));
+            }
         }
     }
 }

--- a/src/Cake.Core.Tests/Unit/Scripting/ScriptTests.cs
+++ b/src/Cake.Core.Tests/Unit/Scripting/ScriptTests.cs
@@ -15,7 +15,7 @@ namespace Cake.Core.Tests.Unit.Scripting
             public void Should_Not_Throw_If_Namespaces_Are_Null()
             {
                 // Given, When
-                var script = new Script(null, new string[] { }, new ScriptAlias[] { }, new string[] { });
+                var script = new Script(null, new string[] { }, new ScriptAlias[] { }, new string[] { }, new string[] { });
 
                 // Then
                 Assert.Equal(0, script.Namespaces.Count);
@@ -25,7 +25,7 @@ namespace Cake.Core.Tests.Unit.Scripting
             public void Should_Not_Throw_If_Lines_Are_Null()
             {
                 // Given, When
-                var script = new Script(new string[] { }, null, new ScriptAlias[] { }, new string[] { });
+                var script = new Script(new string[] { }, null, new ScriptAlias[] { }, new string[] { }, new string[] { });
 
                 // Then
                 Assert.Equal(0, script.Lines.Count);
@@ -35,7 +35,7 @@ namespace Cake.Core.Tests.Unit.Scripting
             public void Should_Not_Throw_If_Aliases_Are_Null()
             {
                 // Given, When
-                var script = new Script(new string[] { }, new string[] { }, null, new string[] { });
+                var script = new Script(new string[] { }, new string[] { }, null, new string[] { }, new string[] { });
 
                 // Then
                 Assert.Equal(0, script.Aliases.Count);
@@ -45,10 +45,20 @@ namespace Cake.Core.Tests.Unit.Scripting
             public void Should_Not_Throw_If_Using_Alias_Directives_Are_Null()
             {
                 // Given, When
-                var script = new Script(new string[] { }, new string[] { }, new ScriptAlias[] { }, null);
+                var script = new Script(new string[] { }, new string[] { }, new ScriptAlias[] { }, null, new string[] { });
 
                 // Then
                 Assert.Equal(0, script.UsingAliasDirectives.Count);
+            }
+
+            [Fact]
+            public void Should_Not_Throw_If_Using_Static_Directives_Are_Null()
+            {
+                // Given, When
+                var script = new Script(new string[] { }, new string[] { }, new ScriptAlias[] { }, new string[] { }, null);
+
+                // Then
+                Assert.Equal(0, script.UsingStaticDirectives.Count);
             }
         }
     }

--- a/src/Cake.Core/Scripting/Analysis/IScriptInformation.cs
+++ b/src/Cake.Core/Scripting/Analysis/IScriptInformation.cs
@@ -44,6 +44,12 @@ namespace Cake.Core.Scripting.Analysis
         IList<string> UsingAliases { get; }
 
         /// <summary>
+        /// Gets all types referenced with the using static directive.
+        /// </summary>
+        /// <value>The fully qualified type names referenced by using static in the script.</value>
+        IList<string> UsingStaticDirectives { get; }
+
+        /// <summary>
         /// Gets the tools.
         /// </summary>
         /// <value>The tools.</value>

--- a/src/Cake.Core/Scripting/Analysis/ScriptAnalyzerResult.cs
+++ b/src/Cake.Core/Scripting/Analysis/ScriptAnalyzerResult.cs
@@ -16,6 +16,7 @@ namespace Cake.Core.Scripting.Analysis
         private readonly HashSet<string> _references;
         private readonly HashSet<string> _namespaces;
         private readonly HashSet<string> _usingAliases;
+        private readonly HashSet<string> _usingStaticDirectives;
 
         /// <summary>
         /// Gets the analyzed script.
@@ -46,6 +47,11 @@ namespace Cake.Core.Scripting.Analysis
         /// </summary>
         /// <value>The using aliases.</value>
         public ISet<string> UsingAliases => _usingAliases;
+
+        /// <summary>
+        /// Gets all types referenced with the using static directive.
+        /// </summary>
+        public ISet<string> UsingStaticDirectives => _usingStaticDirectives;
 
         /// <summary>
         /// Gets the addins.
@@ -82,6 +88,7 @@ namespace Cake.Core.Scripting.Analysis
             _references = new HashSet<string>(Collect(script, i => i.References));
             _namespaces = new HashSet<string>(Collect(script, i => i.Namespaces));
             _usingAliases = new HashSet<string>(Collect(script, i => i.UsingAliases));
+            _usingStaticDirectives = new HashSet<string>(Collect(script, i => i.UsingStaticDirectives));
             Tools = new HashSet<PackageReference>(Collect(script, i => i.Tools));
             Addins = new HashSet<PackageReference>(Collect(script, i => i.Addins));
             Errors = errors ?? new List<ScriptAnalyzerError>(0);

--- a/src/Cake.Core/Scripting/Analysis/ScriptInformation.cs
+++ b/src/Cake.Core/Scripting/Analysis/ScriptInformation.cs
@@ -14,6 +14,7 @@ namespace Cake.Core.Scripting.Analysis
         private readonly List<string> _references;
         private readonly List<string> _usingAliases;
         private readonly List<string> _namespaces;
+        private readonly List<string> _usingStaticDirectives;
         private readonly List<PackageReference> _addins;
         private readonly List<PackageReference> _tools;
         private readonly List<string> _lines;
@@ -28,6 +29,8 @@ namespace Cake.Core.Scripting.Analysis
 
         public IList<string> Namespaces => _namespaces;
 
+        public IList<string> UsingStaticDirectives => _usingStaticDirectives;
+
         public IList<PackageReference> Addins => _addins;
 
         public IList<PackageReference> Tools => _tools;
@@ -41,6 +44,7 @@ namespace Cake.Core.Scripting.Analysis
             _references = new List<string>();
             _namespaces = new List<string>();
             _usingAliases = new List<string>();
+            _usingStaticDirectives = new List<string>();
             _addins = new List<PackageReference>();
             _tools = new List<PackageReference>();
             _lines = new List<string>();

--- a/src/Cake.Core/Scripting/Processors/UsingStatementProcessor.cs
+++ b/src/Cake.Core/Scripting/Processors/UsingStatementProcessor.cs
@@ -44,6 +44,13 @@ namespace Cake.Core.Scripting.Processors
                 return true;
             }
 
+            // Using static directive?
+            if (tokens.Length == 3 && tokens[1].Equals("static", StringComparison.Ordinal))
+            {
+                context.Current.UsingStaticDirectives.Add(string.Join(" ", tokens));
+                return true;
+            }
+
             // Namespace
             context.Current.Namespaces.Add(@namespace);
             return true;

--- a/src/Cake.Core/Scripting/Script.cs
+++ b/src/Cake.Core/Scripting/Script.cs
@@ -16,6 +16,7 @@ namespace Cake.Core.Scripting
         private readonly List<string> _lines;
         private readonly List<ScriptAlias> _aliases;
         private readonly List<string> _usingAliasDirectives;
+        private readonly List<string> _usingStaticDirectives;
 
         /// <summary>
         /// Gets the namespaces imported via the <c>using</c> statement.
@@ -44,22 +45,31 @@ namespace Cake.Core.Scripting
         public IReadOnlyList<string> UsingAliasDirectives => _usingAliasDirectives;
 
         /// <summary>
+        /// Gets the using static directives.
+        /// </summary>
+        /// <value>The using static directives.</value>
+        public IReadOnlyList<string> UsingStaticDirectives => _usingStaticDirectives;
+
+        /// <summary>
         /// Initializes a new instance of the <see cref="Script" /> class.
         /// </summary>
         /// <param name="namespaces">The namespaces.</param>
         /// <param name="lines">The scrip lines.</param>
         /// <param name="aliases">The script aliases.</param>
         /// <param name="usingAliasDirectives">The using alias directives.</param>
+        /// <param name="usingStaticDirectives">The using static directives.</param>
         public Script(
             IEnumerable<string> namespaces,
             IEnumerable<string> lines,
             IEnumerable<ScriptAlias> aliases,
-            IEnumerable<string> usingAliasDirectives)
+            IEnumerable<string> usingAliasDirectives,
+            IEnumerable<string> usingStaticDirectives)
         {
             _namespaces = new List<string>(namespaces ?? Enumerable.Empty<string>());
             _lines = new List<string>(lines ?? Enumerable.Empty<string>());
             _aliases = new List<ScriptAlias>(aliases ?? Enumerable.Empty<ScriptAlias>());
             _usingAliasDirectives = new List<string>(usingAliasDirectives ?? Enumerable.Empty<string>());
+            _usingStaticDirectives = new List<string>(usingStaticDirectives ?? Enumerable.Empty<string>());
         }
     }
 }

--- a/src/Cake.Core/Scripting/ScriptRunner.cs
+++ b/src/Cake.Core/Scripting/ScriptRunner.cs
@@ -207,7 +207,7 @@ namespace Cake.Core.Scripting
             }
 
             // Execute the script.
-            var script = new Script(result.Namespaces, result.Lines, aliases, result.UsingAliases);
+            var script = new Script(result.Namespaces, result.Lines, aliases, result.UsingAliases, result.UsingStaticDirectives);
             session.Execute(script);
         }
 

--- a/src/Cake/Cake.csproj
+++ b/src/Cake/Cake.csproj
@@ -19,7 +19,7 @@
 
   <!-- Global packages -->
   <ItemGroup>
-    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Scripting" Version="2.3.1" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Scripting" Version="2.3.2" />
   </ItemGroup>
 
   <!-- .NET Core packages -->

--- a/src/Cake/Scripting/Roslyn/RoslynCodeGenerator.cs
+++ b/src/Cake/Scripting/Roslyn/RoslynCodeGenerator.cs
@@ -12,10 +12,11 @@ namespace Cake.Scripting.Roslyn
     {
         public string Generate(Script script)
         {
+            var usingStaticDirectives = string.Join("\r\n", script.UsingStaticDirectives);
             var usingDirectives = string.Join("\r\n", script.UsingAliasDirectives);
             var aliases = GetAliasCode(script);
             var code = string.Join("\r\n", script.Lines);
-            return string.Join("\r\n", usingDirectives, aliases, code);
+            return string.Join("\r\n", usingStaticDirectives, usingDirectives, aliases, code);
         }
 
         private static string GetAliasCode(Script context)

--- a/tests/integration/Cake.Core/Scripting/UsingDirective.cake
+++ b/tests/integration/Cake.Core/Scripting/UsingDirective.cake
@@ -1,0 +1,12 @@
+using static System.Math;
+
+Task("Cake.Core.Scripting.UsingDirective.UsingStatic")
+    .Does(() =>
+{
+    Information(Round(1.1));
+});
+
+//////////////////////////////////////////////////////////////////////////////
+
+Task("Cake.Core.Scripting.UsingDirective")
+    .IsDependentOn("Cake.Core.Scripting.UsingDirective.UsingStatic");

--- a/tests/integration/build.cake
+++ b/tests/integration/build.cake
@@ -21,6 +21,7 @@
 #load "./Cake.Common/Tools/DotNetCore/DotNetCoreAliases.cake"
 #load "./Cake.Common/Tools/NuGet/NuGetAliases.cake"
 #load "./Cake.Core/Scripting/LoadDirective.cake"
+#load "./Cake.Core/Scripting/UsingDirective.cake"
 #load "./Cake.Core/Tooling/ToolLocator.cake"
 #load "./Cake.Core/CakeAliases.cake"
 
@@ -45,6 +46,7 @@ Setup(ctx =>
 
 Task("Cake.Core")
     .IsDependentOn("Cake.Core.Scripting.LoadDirective")
+    .IsDependentOn("Cake.Core.Scripting.UsingDirective")
     .IsDependentOn("Cake.Core.Tooling.ToolLocator")
     .IsDependentOn("Cake.Core.CakeAliases");
 

